drivers/can/can_mcp251xfd.c

```diff
@@ -0,0 +1,1814 @@
+		return -EINVAL;
+	}
+
+	reg = mcp251xfd_get_spi_buf_ptr(dev);
+
+	if (is_enabled) {
+		*reg = FIELD_PREP(MCP251XFD_REG_TDC_TDCMOD_MASK, MCP251XFD_REG_TDC_TDCMOD_AUTO);
+		*reg |= FIELD_PREP(MCP251XFD_REG_TDC_TDCO_MASK, tdc_offset);
+	} else {
+		*reg = FIELD_PREP(MCP251XFD_REG_TDC_TDCMOD_MASK, MCP251XFD_REG_TDC_TDCMOD_DISABLED);
+	}
+
+	*reg = sys_cpu_to_le32(*reg);
+
+	return mcp251xfd_write(dev, MCP251XFD_REG_TDC, MCP251XFD_REG_SIZE);
+}
+
+static int mcp251xfd_set_mode_internal(const struct device *dev, uint8_t requested_mode)
+{
+	struct mcp251xfd_data *dev_data = dev->data;
+	uint32_t *reg;
+	uint32_t opmod, reg_con;
+	int ret = 0;
+
+	k_mutex_lock(&dev_data->mutex, K_FOREVER);
+
+	reg = mcp251xfd_read_crc(dev, MCP251XFD_REG_CON, MCP251XFD_REG_SIZE);
+	if (!reg) {
+		ret = -EINVAL;
+		goto done;
+	}
+
+	reg_con = sys_le32_to_cpu(*reg);
+
+	opmod = FIELD_GET(MCP251XFD_REG_CON_OPMOD_MASK, reg_con);
+	if (opmod == requested_mode) {
+		goto done;
+	}
+
+#if defined(CONFIG_CAN_FD_MODE)
+	if (dev_data->current_mcp251xfd_mode == MCP251XFD_REG_CON_MODE_CONFIG) {
+		if (requested_mode ==  MCP251XFD_REG_CON_MODE_CAN2_0 ||
+		    requested_mode ==  MCP251XFD_REG_CON_MODE_EXT_LOOPBACK ||
+		    requested_mode == MCP251XFD_REG_CON_MODE_INT_LOOPBACK) {
+			ret = mcp251xfd_set_tdc(dev, false, 0);
+		} else if (requested_mode == MCP251XFD_REG_CON_MODE_MIXED) {
+			ret = mcp251xfd_set_tdc(dev, true, dev_data->tdco);
+		}
+
+		if (ret < 0) {
+			goto done;
+		}
+	}
+#endif
+
+	reg_con &= ~MCP251XFD_REG_CON_REQOP_MASK;
+	reg_con |= FIELD_PREP(MCP251XFD_REG_CON_REQOP_MASK, requested_mode);
+
+	*reg = sys_cpu_to_le32(reg_con);
+
+	ret = mcp251xfd_write(dev, MCP251XFD_REG_CON, MCP251XFD_REG_SIZE);
+	if (ret < 0) {
+		LOG_ERR("Failed to write REG_CON register [%d]", MCP251XFD_REG_CON);
+		goto done;
+	}
+
+	ret = mcp251xfd_reg_check_value_wtimeout(
+		dev, MCP251XFD_REG_CON, FIELD_PREP(MCP251XFD_REG_CON_OPMOD_MASK, requested_mode),
+		MCP251XFD_REG_CON_OPMOD_MASK, MCP251XFD_MODE_CHANGE_TIMEOUT_USEC,
+		MCP251XFD_MODE_CHANGE_RETRIES, true);
+done:
+	k_mutex_unlock(&dev_data->mutex);
+	return ret;
+}
+
+static int mcp251xfd_set_mode(const struct device *dev, can_mode_t mode)
+{
+	struct mcp251xfd_data *dev_data = dev->data;
+
+	if (dev_data->started) {
+		return -EBUSY;
+	}
+
+	/* todo: Add CAN_MODE_ONE_SHOT support */
+	if ((mode & (CAN_MODE_3_SAMPLES | CAN_MODE_ONE_SHOT)) != 0) {
+		return -ENOTSUP;
+	}
+
+	if (mode == CAN_MODE_NORMAL) {
+		dev_data->next_mcp251xfd_mode = MCP251XFD_REG_CON_MODE_CAN2_0;
+	}
+
+	if ((mode & CAN_MODE_FD) != 0) {
+#if defined(CONFIG_CAN_FD_MODE)
+		dev_data->next_mcp251xfd_mode = MCP251XFD_REG_CON_MODE_MIXED;
+#else
+		return -ENOTSUP;
+#endif
+	}
+
+	if ((mode & CAN_MODE_LISTENONLY) != 0) {
+		dev_data->next_mcp251xfd_mode = MCP251XFD_REG_CON_MODE_LISTENONLY;
+	}
+
+	if ((mode & CAN_MODE_LOOPBACK) != 0) {
+		dev_data->next_mcp251xfd_mode = MCP251XFD_REG_CON_MODE_EXT_LOOPBACK;
+	}
```

Comment:
These lines will overwrite `dev_data->next_mcp251xfd_mode` - does this imply, that e.g. CAN-FD and external loopback cannot be enabled simultaneously?

Reply:
Yes, that's right. Loopback is it's own separate mode:

![image](https://github.com/zephyrproject-rtos/zephyr/assets/17433307/c30e928a-7840-4b6a-8264-04101c61b6a5)

Reply:
So CAN-FD and loopback mode are mutually exclusive? Do the tests in `tests/drivers/can/api` pass?

Reply:
Yes, all the tests pass. There are three that are skipped:

 - SKIP - [can_classic.test_recover] duration = 0.001 seconds  
 - SKIP - [can_classic.test_send_receive_ext_id_rtr] duration = 0.001 seconds                                                 
 - SKIP - [can_classic.test_send_receive_std_id_rtr] duration = 0.001 seconds

Reply:
And the CAN-FD tests pass as well?

Reply:
Yes, all looks fine:
https://pastebin.com/a5SSSY5M